waku/v2/protocol/waku_filter_v2/protocol.nim

```diff
@@ -0,0 +1,159 @@
+## Waku Filter protocol for subscribing and filtering messages
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/[options,sequtils,sets,tables],
+  chronicles,
+  chronos,
+  libp2p/peerid,
+  libp2p/protocols/protocol
+import
+  ../../node/peer_manager,
+  ../waku_message,
+  ./common,
+  ./protocol_metrics,
+  ./rpc_codec,
+  ./rpc
+
+logScope:
+  topics = "waku filter"
+
+const
+  MaxSubscriptions* = 1000 # TODO make configurable
+  MaxCriteriaPerSubscription = 1000
+
+type
+  FilterCriterion* = (PubsubTopic, ContentTopic) # a single filter criterion is fully defined by a pubsub topic and content topic
+  FilterCriteria* = HashSet[FilterCriterion] # a sequence of filter criteria
+
+  WakuFilter* = ref object of LPProtocol
+    subscriptions*: Table[PeerID, FilterCriteria] # a mapping of peer ids to a sequence of filter criteria
+    peerManager: PeerManager
+
+proc pingSubscriber(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  trace "pinging subscriber", peerId=peerId
+
+  if peerId notin wf.subscriptions:
+    debug "pinging peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError.notFound())
+
+  ok()
+
+proc subscribe(wf: WakuFilter, peerId: PeerID, pubsubTopic: Option[PubsubTopic], contentTopics: seq[ContentTopic]): FilterSubscribeResult =
```

Comment:
any reason why `pubsubTopic` is optional?

Reply:
It's an optional in the request RPC since it's only mandatory when following the SUBSCRIBE workflow. I made the choice to do all request consistency checking within the utility function belonging to that request. In this case it is mandatory to have a `pubsubTopic`, so `subscribe` returns a bad request error if optional is none.